Meeting Query Algorithm

lolfuljames opened the pull request:
Current algorithm implementation: 
1. Filter edge cases
2. Filter events based on meeting attendees
3. Combine events if overlap/consecutive
4. Generate vacant slots
5. Filter slots based on meeting duration requirements

UT results:
![image](https://user-images.githubusercontent.com/30533810/85766404-9647b980-b749-11ea-937c-631fc211af1b.png)

lolfuljames committed:
Add Initial Algorithm Implementation for Meeting Schedule Query

This commit is the first implementation for the following algorithm:
	1. Combine timeslots if possible
	2. Create vacant timeslots based on occupied timeslots
	3. Return vacant timeslots

Issues: Query returns empty list

lolfuljames committed:
Modify Algorithm for Scheduling

This commit modifies the algorithm in the following ways:
 - Add dummy events for StartOfDay and EndOfDay

This change makes handling of vacant slots be more streamlined and
graceful.

Issues:
 - Events are unfiltered, need to implement filter based on attendees

lolfuljames committed:
Add Duration Filter to Algorithm

This commit implements a duration filter for vacant timeslots

lolfuljames committed:
Add Event Attendees Filters to Algorithm

This commit:
 - Adds a filter that checks whether events are relevant to
   meeting attendees.
 - Modifies vacant slot duration filter
 - General code cleanup

lolfuljames commented:
Current implementation's bottleneck is probably step 2 - filtering events based on attendees, would like some thoughts on whether it can/should be optimized.

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,108 @@
+  private Boolean isEventIrrelevant(Event event, MeetingRequest meeting) {
+    Collection<String> eventAttendees = event.getAttendees();
+    Collection<String> requiredAttendees = meeting.getAttendees();
+    for (String eventAttendee: eventAttendees) {
```
May consider using `Collections.disjoint()`

lolfuljames replied:
Nice, I had never heard of this method, thanks!

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,108 @@
+    Iterator<TimeRange> vacantSlotsIter = vacantSlots.iterator();
+    // Filter based on duration
+    vacantSlots.removeIf(timeSlot -> isSlotInsufficient(timeSlot, request));
+    return (Collection<TimeRange>)vacantSlots;
```
Unnecessary casting to Collection\<TimeRange>. List<TimeRange> is a Collection\<TimeRange>.

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,108 @@
+    List<TimeRange> occupiedQueue = new ArrayList<TimeRange>();
+    List<Event> events = new ArrayList<Event>();
+    // duplicate events list to support removal of events
+    for (Event event: eventsSource) {
```
Can be replaced with `events.addAll(eventsSource)`.
Or even when initiating the `events` itself using `new ArrayList<>(eventsSource)`.

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,108 @@
+    // handle edge-case entries
+    if (request.getAttendees().size() == 0) {
+      vacantSlots.add(TimeRange.WHOLE_DAY);
+      return vacantSlots;
```
To make the code easier to understand, I suggest using `return Collections.singleton(TimeRange.WHOLE_DAY)`.
Either `singleton` or `singletonList` if `List` is important.

lolfuljames replied:
singletonList will be used as the test expects a list

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,108 @@
+      return vacantSlots;
+    }
+    if (request.getDuration() > TimeRange.END_OF_DAY) {
+      return vacantSlots;
```
To make the code easier to understand, I suggest using `Collections.emptyList()`.

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,108 @@
+      vacantSlots.add(TimeRange.WHOLE_DAY);
+      return vacantSlots;
+    }
+    if (request.getDuration() > TimeRange.END_OF_DAY) {
```
Sometimes it makes it easier to understand the code by extracting the if condition as a method, e.g.
```java
if (isMeetingRequestTooLong(request))
```

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,108 @@
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+  private Boolean isEventIrrelevant(Event event, MeetingRequest meeting) {
```
I think the name of the method can be better. How is it that same attendees make it "irrelevant"?

lolfuljames replied:
Not having any common attendees would make the event irrelevant to the scheduling of the meeting, hence the name.

henryken replied:
I would suggest to name it to explain that it tries to check for common attendees.
"Irrelevant" is too generic and doesn't explain much about the code.

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,108 @@
+      vacantSlots.add(vacantSlot);
+    }
+
+    Iterator<TimeRange> vacantSlotsIter = vacantSlots.iterator();
```
This variable seems not used?

lolfuljames replied:
Yes this was only used in my previous implementation for filtering durations, removed.

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,108 @@
+    }
+    // create events to fill in the gap
+    TimeRange vacantSlot = null;
+    for (int currentEventIdx = 0;
```
It's more common to `i` as index rather than this longer name, unless it is important.

henryken replied:
I suggest to extract methods as necessary for the for loop in order to understand the code better and make it more readable

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,108 @@
+    Collections.sort(occupiedQueue, TimeRange.ORDER_BY_START);
+
+    // combine events to make bigger chunks
+    for (int currentEventIdx = 0;
```
I suggest to extract methods as necessary for the for loop in order to understand the code better and make it more readable

henryken reviewed (commented):
Great work for making the tests pass!
Would be a good exercise to refactor the code to make it more readable and easier to understand.
You could further investigate more ways to make it more optimized afterwards.

lolfuljames committed:
Refactor Meeting Query Code Structure

This commit:
 - Adds method extraction
 - Refactor 'while' loop found during combination of timeslots.

lolfuljames committed:
Rename isEventIrrelevent to hasNoCommonAttendees

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,137 @@
+    // Duplicate events list to support removal of events
+    List<Event> events = new ArrayList<>(eventsSource);
+
+    if (hasNoAttendees(request)) {
```
These two if-blocks can be executed first. The key is to fail fast without having to initialize the `List` above unnecessarily.

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,137 @@
+    }
+
+    // Filter entries that required attendees are not joining
+    filterEvents(events, request);
```
I'd prefer the `.removeIf` here. It's actually more self-explanatory than the generic `filterEvents` name.
Also try to avoid needing to write comments by using more descriptive naming.

Alternatively, you can also rename the method to be more descriptive.

lolfuljames replied:
The idea behind having a generic filterEvents function was to have it act as a wrapper function to support any possible further filtering in the future. However, I agree that the removeIf was a better/more descriptive way of writing it. Changed to removeIf

henryken replied:
The future scenario may not come at all. Name it according to the current intent.
You can always refactor it when necessary.

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,137 @@
+    filterEvents(events, request);
+
+    // Create dummy entries for StartOfDay and EndOfDay
+    initializeOccupiedQueue(occupiedQueue);
```
I'd prefer to have it like the following
```java
List<TimeRange> occupiedQueue = initializeOccupiedQueue();
```

Or you could also name the method `createInitialOccupiedQueue()`.

henryken commented on walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -13,11 +13,137 @@
+    }
+
+    // Filter out vacant slots that do not meet duration requirement
+    filterVacantSlots(vacantSlots, request);
```
I'd prefer the .removeIf here. It's actually more self-explanatory than the generic filterVacantSlots name.
Also try to avoid needing to write comments by using more descriptive naming.

Alternatively, you can also rename the method to be more descriptive.

henryken reviewed (commented):
The code looks much better.
I have a few more comments, just to push you more.

Afterwards, please share your experience on this exercise.

lolfuljames committed:
Refactor Code Structure

This commit:
 - Removes method extraction for event/vacantSlot filtering
 - Removes excess comments

henryken reviewed (approved):
LGTM. Thanks for the refactoring.
Please share your experience working on this PR.

lolfuljames commented:
Thanks for your review as well, this week has been pretty fun! I would like to know if you have any possible optimisations in mind for the performance of the algorithm too!

henryken commented:
Glad that you enjoyed it! I don't know the optimal solution to the problem as I haven't worked on it before.
What we'll do in the next team meeting is to share everyone's algorithm and discuss different solutions to the problem.
